server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java

```diff
@@ -0,0 +1,114 @@
+ */
+
+package io.crate.expression.scalar;
+
+import io.crate.data.Input;
+import io.crate.expression.symbol.Function;
+import io.crate.expression.symbol.Literal;
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+
+public abstract class ConcatwsFunction extends Scalar<String, String> {
+
+    public static final String NAME = "concat_ws";
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.STRING.getTypeSignature()
+            ).withVariableArity(),
+            StringConcatwsFunction::new
+        );
+
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    ConcatwsFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public Signature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    public Symbol normalizeSymbol(Function function, TransactionContext txnCtx, NodeContext nodeCtx) {
+        if (anyNonLiterals(function.arguments())) {
+            return function;
+        }
+        Input[] inputs = new Input[function.arguments().size()];
+        for (int i = 0; i < function.arguments().size(); i++) {
+            inputs[i] = ((Input) function.arguments().get(i));
+        }
+        //noinspection unchecked
+        return Literal.ofUnchecked(boundSignature.getReturnType().createType(), evaluate(txnCtx, nodeCtx, inputs));
+    }
+
+    static class StringConcatwsFunction extends ConcatwsFunction {
+
+        StringConcatwsFunction(Signature signature, Signature boundSignature) {
+            super(signature, boundSignature);
+        }
+
+        @Override
+        public String evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input[] args) {
+
+            if (args[0].value() == null) {
+                throw new NullPointerException("Delimiting argument of the 'concat_ws' function cannot be null.");
+            }
+
+            String delimitArg = (String) args[0].value();
+
+            StringBuilder sb = new StringBuilder();
+
+            for (int i = 1; i < args.length; i++) {
+                String value = (String) args[i].value();
+                if (value != null) {
+                    sb.append(value);
+                    if (i != args.length - 1 && delimitArg != null) {
+                        sb.append(delimitArg);
+                    }
+                }
+            }
+            //In the case the last argument is NULL
+            if (sb.toString().endsWith(",")) {
```

Comment:
I think we can avoid `endsWith/substring` calls for every row in the following way (roughly):
```
boolean firstNonNullProcessed = false;
for (...) {
    if (val != null) {
        if (!firstNonNullProcessed) {
            firstNonNullProcessed = true;
            sb.append(value);
        } else {
             sb.append(separator).append(value).
        }
    }    
}
```

Reply:
Thank you for you comment. This is cleaner. I've implemented your suggested change.